testing/unittests/jthreadpooltests.cpp

```diff
@@ -0,0 +1,943 @@
+
+        duration = measureThreadStartDuration(startTime, 1);
+        CPPUNIT_ASSERT(duration.count() >= 90 && duration.count() < 500);
+        CPPUNIT_ASSERT_EQUAL(4U, threadStartedCount.load());
+        CPPUNIT_ASSERT_EQUAL(3U, pool->runningCount());
+
+        // Test: No throttle delay is introduced after pool falls below defaultmax
+        while (pool->runningCount())
+            std::this_thread::sleep_for(std::chrono::milliseconds(50));
+
+        startTime = std::chrono::high_resolution_clock::now();
+        params = {lifespan1000milliseconds, "Thread7"};
+        pool->start(&params, params.name);
+        params = {lifespan1000milliseconds, "Thread8"};
+        pool->start(&params, params.name);
+        params = {lifespan1000milliseconds, "Thread9"};
+        pool->start(&params, params.name);
+
+        // Should start immediately without throttling
+        duration = measureThreadStartDuration(startTime, 2);
+        CPPUNIT_ASSERT(duration.count() < 100);
+
+        // Test: Pool was above, fell below, and is now at defaultmax again
+        CPPUNIT_ASSERT_EQUAL(3U, pool->runningCount());
+
+        // All threads should complete
+        pool->joinAll(true);
+        CPPUNIT_ASSERT_EQUAL(7U, threadStartedCount.load());
+        CPPUNIT_ASSERT_EQUAL(7U, threadCompletedCount.load());
+        CPPUNIT_ASSERT_EQUAL(0U, pool->runningCount());
+        CPPUNIT_ASSERT(pool->running());
+    }
+
+    void testThrottledThreadPoolWithFastThreadCompletion()
+    {
+        // Test the behavior of a throttled thread pool with 100 default max threads
+        // and 1ms default start delay under high load with fast-completing threads
+        pool.setown(createThreadPool("TestPoolWithFastThreadCompletion", factory, true, nullptr, 100, 1));
+
+        for (unsigned numIterations = 1; numIterations <= 10; numIterations++)
+        {
+            threadStartedCount = 0;
+            threadCompletedCount = 0;
+
+            unsigned threadId{0};
+            StringBuffer threadName;
+            auto startTime = std::chrono::high_resolution_clock::now();
+            for (unsigned threadId = 1; threadId <= 200; threadId++)
+            {
+                threadName.appendf("Thread%u", threadId);
+                if (threadId < 101)
+                    params = {lifespan500milliseconds, threadName.str()};
+                else
+                    params = {lifespan1000milliseconds, threadName.str()};
+                pool->start(&params, params.name);
+            }
+
+            auto duration = measureThreadStartDuration(startTime, 200);
+            CPPUNIT_ASSERT(duration.count() < 500);
+            CPPUNIT_ASSERT_EQUAL(200U, pool->runningCount());
+            CPPUNIT_ASSERT_EQUAL(200U, threadStartedCount.load());
+            CPPUNIT_ASSERT_EQUAL(0U, threadCompletedCount.load());
+
+            // Wait for 100 threads to complete, leaving the pool at capacity
+            while (pool->runningCount() > 100)
```

Comment:
this isn't testing if 100 are complete... ? and 100 should be running almost immediately.